modules/layouts/LayoutCS.chpl

```diff
@@ -0,0 +1,656 @@
+/*
+ * Copyright 2004-2017 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/* Debug flag */
+config param debugCS = false;
+use RangeChunk only ;
+
+/* For backwards-compatibility with LayoutCSR */
+type CSR = CS;
+
+pragma "no doc"
+/* Comparator used for sorting by columns */
+record _ColumnComparator {
+  proc init() { }
+  proc key(idx: _tuple) { return (idx(2), idx(1));}
+}
+
+const _columnComparator: _ColumnComparator;
+
+/*
+This CS layout provides a Compressed Sparse Row and Compressed Sparse Column
+implementation for Chapel's sparse domains and arrays.
+
+To declare a CS domain, invoke the ``CS`` constructor without arguments
+in a `dmapped` clause. For example:
+
+  .. code-block:: chapel
+
+    use LayoutCS;
+    var D = {1..n, 1..m};  // a default-distributed domain
+    var CS_Domain: sparse subdomain(D) dmapped CS();
+
+To declare a CS array, use a CS domain, for example:
+
+  .. code-block:: chapel
+
+    // assume the above declarations
+    var CS_Array: [CS_Domain] real;
+
+This domain map is a layout, i.e. it maps all indices to the current locale.
+All elements of a CS-distributed array are stored
+on the locale where the array variable is declared.
+*/
+class CS: BaseDist {
+  const row: bool = true;
+
+  proc dsiNewSparseDom(param rank: int, type idxType, dom: domain) {
+    return new CSDom(rank=rank, row=row, idxType=idxType, dist=this, parentDom=dom);
+  }
+
+  proc dsiClone() return new CS(row=this.row);
+
+  proc dsiEqualDMaps(that: CS) param {
+    return true;
+  }
+
+  proc dsiEqualDMaps(that) param {
+    return false;
+  }
+} // CS
+
+
+class CSDom: BaseSparseDomImpl {
+  const row: bool;                  // would like param
+
+  var dist: CS;
+
+  var rowRange: range(idxType);
+  var colRange: range(idxType);
+
+  /* (row|col) dom */
+  const dom: domain(1, idxType);
+
+  /* (row|col) start */
+  var startIdx: [dom] idxType;      // would like index(nnzDom)
+  /* (row|col) idx */
+  var idx: [nnzDom] idxType;        // would like index(parentDom.dim(1))
+
+  /* Initializer */
+  proc CSDom(param rank, type idxType, const row=true,
```

Comment:
I don't think there's a code path that would ever use the default value of `row` here, so I'd probably remove it for simplicity.